lib/cn/kvcompact.c

```diff
@@ -200,10 +200,11 @@ cn_kvcompact(struct cn_compaction_work *w)
                                       &vboff, &vdata, &vlen, &complen))
                 break;
 
-            omlen = (vtype == vtype_val) ? vlen : ((vtype == vtype_cval) ? complen : 0);
-
-            direct = omlen > direct_read_len;
-            if (direct) {
+            /* Compute on-media length (w/o branches) to see if direct vblock read should
+             * be used. Note omlen will be 0 for vtypes that are not stored in vblocks.
+             */
+            omlen = (vtype == vtype_val) * vlen + (vtype == vtype_cval) * complen;
+            if (omlen > direct_read_len) {
```

Comment:
Well, I suppose it might have generated cmov or something, seems like for sure we now have an imult, so not obvious it's faster/better?

Reply:
Original:
```
00000000000001c0 <omlen_old>:
     1c0:       41 54                   push   %r12
     1c2:       41 89 f4                mov    %esi,%r12d
     1c5:       85 ff                   test   %edi,%edi
     1c7:       74 0a                   je     1d3 <omlen_old+0x13>
     1c9:       45 31 e4                xor    %r12d,%r12d
     1cc:       83 ff 05                cmp    $0x5,%edi
     1cf:       44 0f 44 e2             cmove  %edx,%r12d
     1d3:       44 89 e6                mov    %r12d,%esi
     1d6:       48 8d 3d 00 00 00 00    lea    0x0(%rip),%rdi        # 1dd <omlen_old+0x1d>
     1dd:       31 c0                   xor    %eax,%eax
     1df:       e8 00 00 00 00          call   1e4 <omlen_old+0x24>
     1e4:       44 89 e0                mov    %r12d,%eax
     1e7:       41 5c                   pop    %r12
     1e9:       c3                      ret    
     1ea:       66 0f 1f 44 00 00       nopw   0x0(%rax,%rax,1)
```

Patch:
```
00000000000001f0 <omlen_new>:
     1f0:       41 54                   push   %r12
     1f2:       45 31 e4                xor    %r12d,%r12d
     1f5:       85 ff                   test   %edi,%edi
     1f7:       41 0f 94 c4             sete   %r12b
     1fb:       31 c0                   xor    %eax,%eax
     1fd:       44 0f af e6             imul   %esi,%r12d
     201:       83 ff 05                cmp    $0x5,%edi
     204:       48 8d 3d 00 00 00 00    lea    0x0(%rip),%rdi        # 20b <omlen_new+0x1b>
     20b:       0f 94 c0                sete   %al
     20e:       0f af c2                imul   %edx,%eax
     211:       41 01 c4                add    %eax,%r12d
     214:       31 c0                   xor    %eax,%eax
     216:       44 89 e6                mov    %r12d,%esi
     219:       e8 00 00 00 00          call   21e <omlen_new+0x2e>
     21e:       44 89 e0                mov    %r12d,%eax
     221:       41 5c                   pop    %r12
     223:       c3                      ret    
     224:       66 66 2e 0f 1f 84 00    data16 cs nopw 0x0(%rax,%rax,1)
     22b:       00 00 00 00 
     22f:       90                      nop
```

Reply:
Better the old way?

Reply:
Hard to know!  If you don't branch orig looks much better...    BTW, I like -Mintel with objdump, can't stand ATT syntax...